### PR TITLE
Make postcodes mandatory and valid

### PR DESCRIPTION
**Pull request by georgeschena**

### CODE REVIEW

File: app/views/locations/new.erb [hunk 1]
```diff
@@ -37,7 +42,11 @@
 
       <div class="govuk-form-group">
         <%= f.label :postcode, class: 'govuk-label' %>
-        <%= f.text_field :postcode, class: 'govuk-input govuk-input--width-10' %>
+        <% if @location.errors.keys.include?(:postcode) %>
+          <%= f.text_field :postcode, class: 'govuk-input govuk-input--error' %>
+        <% else %>
+          <%= f.text_field :postcode, class: 'govuk-input govuk-input--width-10' %>
```
Comment (kjdchapman) [inline app/views/locations/new.erb]: Does this width class need to be on line 46 as well?

File: app/models/location.rb [hunk 2]
```diff
@@ -7,15 +9,23 @@ class Location < ApplicationRecord
   validates_associated :ips
 
   validates :address, presence: true
+  validates :postcode, presence: true
```
Comment (emileswarts) [inline app/models/location.rb]: Think you can squash the above two into a single line

validates :postcode, :address, presence: true

File: app/models/location.rb [hunk 3]
```diff
@@ -7,15 +9,23 @@ class Location < ApplicationRecord
...
 
   def full_address
-    postcode.blank? ? address : "#{address}, #{postcode}"
+    "#{address}, #{postcode}"
   end
 
 private
 
+  def validate_postcode
+    if postcode.nil? || !UKPostcode.parse(postcode).valid?
```
Comment (emileswarts) [inline app/models/location.rb]: Do we need a nil check?
Won't nil be invalid according to this gem?

File: spec/models/location_spec.rb [hunk 4]
```diff
@@ -34,13 +35,19 @@
...
       end
+    end
+  end
 
-      context 'but a blank postcode' do
-        let(:postcode) { '' }
+  describe 'invalid postcode entry' do
+    context 'with an invalid postcode' do
+      it "errors if the postcode is not valid" do
+        subject.postcode = "WHATEVER POSTCODE"
```
Comment (emileswarts) [inline spec/models/location_spec.rb]: I'd stick to single quotes for non interpolated strings for consistency

File: app/models/location.rb (hunk 3 above)
Comment (georgeschena) [reply app/models/location.rb]: In the Location model tests it moans about postcode being nil as on create I assume it doesn't  create a postcode from what I'm aware of? @emileswarts

Comment (emileswarts) [reply app/models/location.rb]: So can the first part of this conditional be removed? Not 100% clear on this.

Comment (georgeschena) [reply app/models/location.rb]: It can't

Comment (emileswarts) [reply app/models/location.rb]:
```ruby
unless UKPostcode.parse(postcode.to_s).valid?
     errors.add(:postcode, "must be valid")
end
```

File: app/models/location.rb [hunk 5]
```diff
@@ -6,16 +8,23 @@ class Location < ApplicationRecord
...
+      errors.add(:postcode, "must be valid")
```
Comment (kjdchapman) [inline app/models/location.rb]: nit: single quotes
```suggestion
      errors.add(:postcode, 'must be valid')
```

File: app/models/location.rb (hunk 3 above)
Comment (georgeschena) [reply app/models/location.rb]: I LIKE

File: spec/features/locations/add_new_location_spec.rb [hunk 6]
```diff
@@ -81,13 +81,37 @@
...
+
+      context 'the postcode is blank' do
+        before do
+          fill_in 'Address', with: '30 Square'
+          fill_in 'Postcode', with: ''
+          fill_in ip_input, with: '10.0.0.3'
+          click_on 'Add new location'
+        end
+
+        it_behaves_like "errors in form"
```
Comment (kjdchapman) [inline spec/features/locations/add_new_location_spec.rb]: nit: quote marks
```suggestion
        it_behaves_like 'errors in form'
```

File: spec/models/location_spec.rb [hunk 7]
```diff
@@ -34,13 +35,24 @@
           expect(subject.full_address).to eq('121 Fictional Street, FI5 S67')
         end
       end
+    end
+  end
+
+  describe 'Entering a postcode' do
+    context 'which is in the incorrect format' do
+      it 'errors as the postcode dosent match the correct format' do
```
Comment (kjdchapman) [inline spec/models/location_spec.rb]: nit: spelling
```suggestion
      it 'errors as the postcode does not match the correct format' do
```

**Pull request merged**